src/main/scala/mesosphere/marathon/core/storage/PersistenceStore.scala

```diff
@@ -0,0 +1,254 @@
+package mesosphere.marathon.core.storage
+
+import akka.http.scaladsl.marshalling.{ Marshal, Marshaller }
+import akka.http.scaladsl.unmarshalling.Unmarshaller
+import akka.stream.Materializer
+import akka.stream.scaladsl.Source
+import akka.{ Done, NotUsed }
+import mesosphere.marathon.StoreCommandFailedException
+import mesosphere.util.CallerThreadExecutionContext
+
+import scala.async.Async._
+import scala.concurrent.{ ExecutionContext, Future }
+import scala.util.{ Failure, Success, Try }
+
+/**
+  * Resolver for Marathon Internal Persistence IDs (by `Key Type`, `Value Type` and `Serialized Type`)
+  * For example, Applications may be identified by the string "/apps/abc" and stored internally at
+  * "/apps/a/a58ec17d-735d-4c3f-9aa8-d44c764aa31b"
+  *
+  * This IdResolver must be provided for all storage engines, e.g.
+  *
+  * {{{
+  *   class A(val id: String) {}
+  *   object A {
+  *     implicit val zkIdResolver = new IdResolver[String, A, ByteString] = {
+  *       def toStorageId(id: String): String = id
+  *       def fromStorageId(id: String): String = id
+  *   }
+  * }}}
+  *
+  * @tparam Id The marathon type of the ID for the given Value type
+  * @tparam K The persisted type of the ID
+  * @tparam V The value type being stored
+  * @tparam Serialized The serialized type of 'V' for the given [[PersistenceStore]].
+  */
+trait IdResolver[Id, K, +V, +Serialized] {
```

Comment:
I see your solution regarding different storage engines: the resolver is defined with respect to a specific key and serialized type - nice!